hledger-lib/Hledger/Data/AccountBalance.hs

```diff
@@ -0,0 +1,159 @@
+    maybe h snd $ IM.lookupLE (fromInteger $ toModifiedJulianDay d) as
+
+-- | Add the 'AccountBalance' to the appropriate location in 'AccountBalances'.
+insertAccountBalances :: Semigroup a => Maybe Day -> a -> AccountBalances a -> AccountBalances a
+insertAccountBalances mday b balances = case mday of
+    Nothing  -> balances{abhistorical = abhistorical balances <> b}
+    Just day -> balances{abdatemap = IM.insertWith (<>) (fromInteger $ toModifiedJulianDay day) b $ abdatemap balances}
+
+-- | Performs an operation on the contents of two 'AccountBalances'.
+--
+-- This will drop keys if they are not present in both AccountBalances.
+opAccountBalances :: (a -> b -> c) -> AccountBalances a -> AccountBalances b -> AccountBalances c
+opAccountBalances f (AccountBalances h1 as1) (AccountBalances h2 as2) =
+    AccountBalances (f h1 h2) $ IM.intersectionWith f as1 as2
+
+-- | Performs an operation on the contents of two 'AccountBalances', with
+-- separate treatments for those only in the first, only in the second, or in
+-- both 'AccountBalance's.
+mergeAccountBalances :: (a -> c) -> (b -> c) -> (a -> b -> c)
+                     -> AccountBalances a -> AccountBalances b -> AccountBalances c
+mergeAccountBalances only1 only2 f = \(AccountBalances h1 as1) (AccountBalances h2 as2) ->
+    AccountBalances (f h1 h2) $ merge as1 as2
+  where
+    merge = IM.mergeWithKey (\_ x y -> Just $ f x y) (fmap only1) (fmap only2)
+
+-- | Pad out the datemap of an 'AccountBalances' so that every key from a set is present.
+padAccountBalances :: Monoid a => IS.IntSet -> AccountBalances a -> AccountBalances a
+padAccountBalances keys bal = bal{abdatemap = abdatemap bal <> IM.fromSet (const mempty) keys}
+
+
+instance Show AccountBalance where
+  showsPrec d (AccountBalance n e i) =
+    showParen (d > 10) $
+        showString "AccountBalance"
+      . showString "{ abnumpostings = " . shows n
+      . showString ", abebalance = " . showString (wbUnpack (showMixedAmountB defaultFmt e))
+      . showString ", abibalance = " . showString (wbUnpack (showMixedAmountB defaultFmt i))
+      . showChar '}'
+
+instance Semigroup AccountBalance where
+  AccountBalance n e i <> AccountBalance n' e' i' = AccountBalance (n + n') (maPlus e e') (maPlus i i')
+
+instance Monoid AccountBalance where
+  mempty = AccountBalance 0 nullmixedamt nullmixedamt
+
+-- | Apply an operation to both 'MixedAmount' in an 'AccountBalance'.
+applyAccountBalance :: (MixedAmount -> MixedAmount) -> AccountBalance -> AccountBalance
+applyAccountBalance f a = a{abebalance = f $ abebalance a, abibalance = f $ abibalance a}
+
+-- | Perform an operation on the 'MixedAmount' in two 'AccountBalance'.
+opAccountBalance :: (MixedAmount -> MixedAmount -> MixedAmount) -> AccountBalance -> AccountBalance -> AccountBalance
+opAccountBalance f a b = a{abebalance = f (abebalance a) (abebalance b), abibalance = f (abibalance a) (abibalance b)}
+
+
+-- tests
+
+tests_AccountBalance = testGroup "AccountBalance" [
```

Comment:
Keep these separate for clarity if you like, but I suggest grouping them under tests_AccountBalance for export to avoid breaking the consistent pattern (every module exports one similarly named test group)

Reply:
I'll follow your later advice and split these apart into separate modules (`AccountBalance` and `AccountBalances`).